### PR TITLE
`Learning path`: Add discussion section to lecture unit and exercises

### PR DESCRIPTION
**Pull request by JohannesWt**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

This PR adds the discussion section to lecture units and exercises in the student's learning path view.

### Description

The discussion section (communication feature) is now displayed alongside lecture units and exercises in the student's learning path view. Instead of the router-based approach previously used in the course details view, the discussion section component is now integrated directly to simplify the implementation.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Students
- at least 1 competency
- 1 Lecture Unit connected to a competency
- 1 Exercise connected to a competency

1. Log in to Artemis
2. enable the communication feature for the specified course (if not all ready enabled)
3. Navigate the learning path student view
4. verify that the discussion section is shown for lecture units (for the specific lecture) as well as for exercises (for the specific exercise)
5. verify that the discussion sections in the student of of lectures and exercises still work as expected

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| learning-path-lecture-unit.component.ts | 97.05% | ✅  |
| learning-path-lecture-unit-view.component.ts | 91.66% | ✅  |
| course-lecture-details.component.ts | 98.61% | ✅  |
| discussion-section.component.ts | 94.69% | ✅  |
| course-exercise-details.component.ts | 90.29% | ✅  |

### Screenshots

image
image

## Summary by CodeRabbit

- **New Features**
	- Enhanced layout and functionality for lecture units, allowing for discussion features based on communication settings.
	- Improved user interaction by integrating discussion capabilities within lecture units.

- **Refactor**
	- Streamlined component interactions and dependencies, particularly regarding discussion functionalities.
	- Transitioned the discussion section component to a standalone entity, enhancing modularity.

- **Tests**
	- Improved test suite for the discussion section, optimizing setup and enhancing clarity.
	- Added new test cases to validate the visibility of discussion sections based on communication settings.
	- Updated tests for the CourseExerciseDetailsComponent to reflect changes in component dependencies and communication settings.

### CODE REVIEW

File: src/test/javascript/spec/component/overview/discussion-section/discussion-section.component.spec.ts [hunk 1]
```diff
@@ -89,48 +87,39 @@ describe('DiscussionSectionComponent', () => {
...
+                new HttpResponse({
+                    body: metisExerciseChannelDTO,
+                    status: 200,
+                }),
+            ),
+        );
+        metisServiceGetFilteredPostsSpy = jest.spyOn(metisService, 'getFilteredPosts');
+        component.lecture = { ...metisLecture, course: metisCourse };
+        component.ngOnInit();
+        fixture.detectChanges();
```
Comment (pzdr7) [inline src/test/javascript/spec/component/overview/discussion-section/discussion-section.component.spec.ts:122]: (optional) Can you check if we need both `detectChanges` and `ngOnInit` here?

File: src/main/webapp/app/course/learning-paths/components/learning-path-lecture-unit/learning-path-lecture-unit.component.ts [hunk 2]
```diff
@@ -29,6 +31,11 @@ export class LearningPathLectureUnitComponent {
     readonly isLoading: WritableSignal<boolean> = signal(false);
     readonly lectureUnit = signal<LectureUnit | undefined>(undefined);
 
+    readonly lecture = computed(() => this.lectureUnit()?.lecture);
+
+    readonly isCommunicationEnabled = computed(() => isCommunicationEnabled(this.lecture()?.course));
+    readonly isMessagingEnabled = computed(() => isMessagingEnabled(this.lecture()?.course));
```
Comment (pzdr7) [inline src/main/webapp/app/course/learning-paths/components/learning-path-lecture-unit/learning-path-lecture-unit.component.ts:37]: Looking at the implementation of `isCommunicationEnabled` and `isMessagingEnabled`, it looks to me like `isCommunicationEnabled` is enough for this view:

```
export function isCommunicationEnabled(course: Course | undefined) {
    const config = course?.courseInformationSharingConfiguration;
    return config === CourseInformationSharingConfiguration.COMMUNICATION_AND_MESSAGING || config === CourseInformationSharingConfiguration.COMMUNICATION_ONLY;
}

export function isMessagingEnabled(course: Course | undefined) {
    const config = course?.courseInformationSharingConfiguration;
    return config === CourseInformationSharingConfiguration.COMMUNICATION_AND_MESSAGING;
}
```

i.e., if `isMessagingEnabled` is true, then `isCommunicationEnabled` is also true.

```suggestion
    readonly isCommunicationEnabled = computed(() => isCommunicationEnabled(this.lecture()?.course));
```

File: src/main/webapp/app/course/learning-paths/components/learning-path-lecture-unit/learning-path-lecture-unit.component.html [hunk 3]
```diff
@@ -6,21 +6,30 @@
...
+                }
+                @case (LectureUnitType.EXERCISE) {
+                    <jhi-exercise-unit [exerciseUnit]="lectureUnit" />
+                }
+                @case (LectureUnitType.ONLINE) {
+                    <jhi-online-unit [lectureUnit]="lectureUnit" (onCompletion)="setLearningObjectCompletion($event)" />
+                }
+            }
+        </div>
+        @if (isCommunicationEnabled() || isMessagingEnabled()) {
```
Comment (pzdr7) [inline src/main/webapp/app/course/learning-paths/components/learning-path-lecture-unit/learning-path-lecture-unit.component.html:29]: Then, you can simplify this condition:

```suggestion
        @if (isCommunicationEnabled()) {
```

**Review by pzdr7 — CHANGES_REQUESTED** [review]
Code looks good 👍 I think getting rid of the outlet makes sense

I found one possible simplification & have one more optional comment

File: src/test/javascript/spec/component/overview/course-lectures/course-lecture-details.component.spec.ts [hunk 4]
```diff
@@ -136,20 +143,22 @@ describe('CourseLectureDetailsComponent', () => {
...
+
+        lectureService = TestBed.inject(LectureService);
+        jest.spyOn(lectureService, 'findWithDetails').mockReturnValue(response);
+        jest.spyOn(lectureService, 'find').mockReturnValue(response);
+
+        fixture = TestBed.createComponent(CourseLectureDetailsComponent);
+        courseLecturesDetailsComponent = fixture.componentInstance;
+        debugElement = fixture.debugElement;
+
+        // mock profileService
```
Comment (az108) [inline src/test/javascript/spec/component/overview/course-lectures/course-lecture-details.component.spec.ts:156]: unnecessary in my opinion

File: src/test/javascript/spec/component/learning-paths/components/learning-path-lecture-unit.component.spec.ts [hunk 5]
```diff
@@ -110,4 +118,25 @@ describe('LearningPathLectureUnitComponent', () => {
...
+        expect(discussionSection).toBeTruthy();
+    });
+
+    it('should not show discussion section when communication is disabled', async () => {
+        const lecture = { ...lectureUnit.lecture, course: { courseInformationSharingConfiguration: CourseInformationSharingConfiguration.DISABLED } };
+        getLectureUnitByIdSpy.mockReturnValue(of({ ...lectureUnit, lecture }));
+
+        fixture.detectChanges();
+        await fixture.whenStable();
+        fixture.detectChanges();
```
Comment (az108) [inline src/test/javascript/spec/component/learning-paths/components/learning-path-lecture-unit.component.spec.ts:137]: why do we need 2 detect changes?

Comment (JohannesWt) [reply src/test/javascript/spec/component/learning-paths/components/learning-path-lecture-unit.component.spec.ts:137]: true here we can omit it but the method before we need it twice.

File: src/test/javascript/spec/component/overview/course-lectures/course-lecture-details.component.spec.ts (hunk 4 above)
Comment (JohannesWt) [reply src/test/javascript/spec/component/overview/course-lectures/course-lecture-details.component.spec.ts:156]: this was just done for convenience I think. This way to get an CSS element you can omit `fixture.debugElement` every time

**Review by az108 — DISMISSED** [review]
Thanks for the quick changes, Code 👍

**Review by florian-glombik — APPROVED** [review]
Tested on ts3, worked well for me

I tested the following:
* sending messages as instructor / student from the exercise/lecture/learning path view
* synchronizing messages between the exercise/learning path view

**Review by florian-glombik — DISMISSED** [review]
Re-tested on ts3 after last changes since https://github.com/ls1intum/Artemis/pull/9264#pullrequestreview-2277089051

**Review by Jan-Thurner — APPROVED** [review]
Tested on TS4. Tested sending messages as a student and as an instructor. both works well. The Provided chat is always the one that belongs to the lecture of the currently displayed competency.

**Review by az108 — DISMISSED** [review]
Thanks for the quick fix, reapprove

**Review by Jan-Thurner — APPROVED** [review]
reapprove after fix for client test

**Review by florian-glombik — DISMISSED** [review]
Re-approve test https://github.com/ls1intum/Artemis/pull/9264#pullrequestreview-2277368230 after fixing client test

**Review by pzdr7 — APPROVED** [review]
Code

**Review by Jan-Thurner — DISMISSED** [review]
reapprove after additional test were added

File: src/test/javascript/spec/component/learning-paths/components/learning-path-lecture-unit.component.spec.ts [hunk 6]
```diff
@@ -109,6 +117,38 @@ describe('LearningPathLectureUnitComponent', () => {
...
 
+    it('should not set current learning object on error', async () => {
+        const completeLectureUnitSpy = jest.spyOn(lectureUnitService, 'completeLectureUnit').mockImplementationOnce(() => {});
+
+        fixture.detectChanges();
+        await fixture.whenStable();
+
+        component.setLearningObjectCompletion({ completed: true, lectureUnit: lectureUnit });
+
+        expect(completeLectureUnitSpy).toHaveBeenCalledWith(lectureUnit.lecture, {
```
Comment (pzdr7) [inline src/test/javascript/spec/component/learning-paths/components/learning-path-lecture-unit.component.spec.ts:128]: Can you check if we can be more specific here?

```suggestion
        expect(completeLectureUnitSpy).toHaveBeenCalledExactlyOnceWith(lectureUnit.lecture, {
```

File: src/test/javascript/spec/component/learning-paths/components/learning-path-lecture-unit.component.spec.ts [hunk 7]
```diff
@@ -109,6 +117,38 @@ describe('LearningPathLectureUnitComponent', () => {
...
+            completed: true,
+            lectureUnit: lectureUnit,
+        });
+        expect(setLearningObjectCompletionSpy).toHaveBeenCalledWith(true);
```
Comment (pzdr7) [inline src/test/javascript/spec/component/learning-paths/components/learning-path-lecture-unit.component.spec.ts:149]: Can you check if we can be more specific here?

```suggestion
        expect(completeLectureUnitSpy).toHaveBeenCalledExactlyOnceWith(lectureUnit.lecture, {
            completed: true,
            lectureUnit: lectureUnit,
        });
        expect(setLearningObjectCompletionSpy).toHaveBeenCalledExactlyOnceWith(true);
```

**Review by az108 — APPROVED** [review]
Added tests and rest of the changes to the code look good to me, reapproval

**Review by Jan-Thurner — APPROVED** [review]

**Review by pzdr7 — APPROVED** [review]

**Review by N0W0RK — APPROVED** [review]
Tested on TS3 lectures aswell as exercises show communication in Learning path view

**Review by florian-glombik — APPROVED** [review]
Re-approve

**Review by MaximilianAnzinger — APPROVED** [review]
Maintainer approved

**Pull request merged**